### ISSUE TITLE
Doesn't work with NPM 6 if also including AJV

### ISSUE THREAD

**Issue by sazzer** [stoplightio/spectral#2003]
**Describe the bug**
When using Spectral in a project building with NPM 6, if AJV is also included then requiring Spectral fails to work correctly.

**To Reproduce**

Given this code:
```
-> % cat package.json
{
  "name": "spectral-ajv-bug",
  "scripts": {
    "start": "node ./start.js"
  },
  "dependencies": {
    "@stoplight/spectral-core": "^1.8.0",
    "@stoplight/spectral-functions": "^1.5.0",
    "@stoplight/spectral-parsers": "^1.0.1",
    "@stoplight/spectral-rulesets": "^1.3.0",
    "ajv": "^8.6.3"
  },
  "devDependencies": {
    "eslint": "^7.6.0"
  }
}

-> % cat start.js
const { Spectral } = require('@stoplight/spectral-core')
```

Running `npm start` gives this:
```
-> % npm start

> spectral-ajv-bug@ start /Users/coxg/source/temp/spectral
> node ./start.js

Error compiling schema, function code: const schema20 = scope.schema[14];const schema21 = scope.schema[15];const schema22 = scope.schema[16];const func0 = scope.func[0];return function validate18(data, {instancePath="", parentData, parentDataProperty, rootData=data}={}){let vErrors = null;let errors = 0;const _errs0 = errors;let valid0 = false;let passing0 = null;const _errs1 = errors;if(!(((((data === -1) || (data === 0)) || (data === 1)) || (data === 2)) || (data === 3))){const err0 = {instancePath,schemaPath:"#/$defs/DiagnosticSeverity/enum",keyword:"enum",params:{allowedValues: schema21.enum},message:"must be equal to one of the allowed values"};if(vErrors === null){vErrors = [err0];}else {vErrors.push(err0);}errors++;}var _valid0 = _errs1 === errors;if(_valid0){valid0 = true;passing0 = 0;}const _errs3 = errors;if(!(((((data === "error") || (data === "warn")) || (data === "info")) || (data === "hint")) || (data === "off"))){const err1 = {instancePath,schemaPath:"#/$defs/HumanReadableSeverity/enum",keyword:"enum",params:{allowedValues: schema22.enum},message:"must be equal to one of the allowed values"};if(vErrors === null){vErrors = [err1];}else {vErrors.push(err1);}errors++;}var _valid0 = _errs3 === errors;if(_valid0 && valid0){valid0 = false;passing0 = [passing0, 1];}else {if(_valid0){valid0 = true;passing0 = 1;}}if(!valid0){const err2 = {instancePath,schemaPath:"#/oneOf",keyword:"oneOf",params:{passingSchemas: passing0},message:"must match exactly one schema in oneOf"};if(vErrors === null){vErrors = [err2];}else {vErrors.push(err2);}errors++;}else {errors = _errs0;if(vErrors !== null){if(_errs0){vErrors.length = _errs0;}else {vErrors = null;}}}if(errors > 0){for(const err3 of vErrors){if((((({"str":"err3"}.keyword !== "errorMessage") && (!{"str":"err3"}.emUsed)) && (({"str":"err3"}.instancePath === instancePath) || (({"str":"err3"}.instancePath.indexOf(instancePath) === 0) && ({"str":"err3"}.instancePath[instancePath.length] === "/")))) && ({"str":"err3"}.schemaPath.indexOf("#") === 0)) && ({"str":"err3"}.schemaPath["#".length] === "/")){{"str":"emErrs0"}.push({"str":"err3"});{"str":"err3"}.emUsed = true;}}if({"str":"emErrs0"}.length){if(vErrors === null){vErrors = [{"str":"err4"}];}else {vErrors.push({"str":"err4"});}errors++;}const emErrs1 = [];for(const err5 of vErrors){if(!{"str":"err5"}.emUsed){{"str":"emErrs1"}.push({"str":"err5"});}}vErrors = emErrs1;errors = {"str":"emErrs1"}.length;}validate18.errors = vErrors;return errors === 0;}
/Users/coxg/source/temp/spectral/node_modules/@stoplight/spectral-core/node_modules/ajv/dist/compile/index.js:121
        throw e;
        ^

SyntaxError: Unexpected token ':'
    at new Function (<anonymous>)
    at Ajv.compileSchema (/Users/coxg/source/temp/spectral/node_modules/@stoplight/spectral-core/node_modules/ajv/dist/compile/index.js:90:30)
    at Ajv.inlineOrCompile (/Users/coxg/source/temp/spectral/node_modules/@stoplight/spectral-core/node_modules/ajv/dist/compile/index.js:149:47)
    at Ajv.resolveRef (/Users/coxg/source/temp/spectral/node_modules/@stoplight/spectral-core/node_modules/ajv/dist/compile/index.js:143:46)
    at Object.code (/Users/coxg/source/temp/spectral/node_modules/@stoplight/spectral-core/node_modules/ajv/dist/vocabularies/core/ref.js:19:47)
    at keywordCode (/Users/coxg/source/temp/spectral/node_modules/@stoplight/spectral-core/node_modules/ajv/dist/compile/validate/index.js:454:13)
    at /Users/coxg/source/temp/spectral/node_modules/@stoplight/spectral-core/node_modules/ajv/dist/compile/validate/index.js:185:25
    at CodeGen.code (/Users/coxg/source/temp/spectral/node_modules/@stoplight/spectral-core/node_modules/ajv/dist/compile/codegen/index.js:439:13)
    at CodeGen.block (/Users/coxg/source/temp/spectral/node_modules/@stoplight/spectral-core/node_modules/ajv/dist/compile/codegen/index.js:568:18)
    at schemaKeywords (/Users/coxg/source/temp/spectral/node_modules/@stoplight/spectral-core/node_modules/ajv/dist/compile/validate/index.js:185:13)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! spectral-ajv-bug@ start: `node ./start.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the spectral-ajv-bug@ start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/coxg/.npm/_logs/2021-12-17T09_48_56_925Z-debug.log
```

**Expected behavior**
No error should occur.

**Environment (remove any that are not applicable):**
 - Node: 14.18.2
 - NPM: 6.14.14
 - OS: macOS 11.6 on arm64

**Additional context**
AJV is included explicitly because eslint depends on an incompatible version. Without the explicit inclusion it sometimes works and sometimes gives a totally different error, depending on NPM module resolution.

```
-> % npm ls ajv
spectral-ajv-bug@ /Users/coxg/source/temp/spectral
├─┬ @stoplight/spectral-core@1.8.0
│ ├── ajv@8.6.3
│ └─┬ ajv-formats@2.1.1
│   └── ajv@8.8.2  deduped
├─┬ @stoplight/spectral-functions@1.5.0
│ └── ajv@8.8.2  deduped
├─┬ @stoplight/spectral-rulesets@1.3.0
│ └── ajv@8.6.3
├── ajv@8.8.2
└─┬ eslint@7.32.0
  ├─┬ @eslint/eslintrc@0.4.3
  │ └── ajv@6.12.6
  ├── ajv@6.12.6
  └─┬ table@6.7.5
    └── ajv@8.8.2  deduped
```

This works perfectly fine under NPM 7, NPM 8 and Yarn, but *not* under NPM 6 - which is what we're currently using.

I get the exact same error whether the explicit dependency on AJV is for 8.6.3 or 8.8.2 - both of which are used by Spectral. I do *not* get any errors if I remove both the AJV and ESLint dependencies from the project.

Note as well that the error occurs by including the `require` line. No functionality from Spectral actually needs to be triggered.

**Comment by P0lip** [comment]
I'm afraid there's not much we can do about it.
You may need to file a bug in Ajv if there isn't any yet.

**Comment by sazzer** [comment]
I raised it here because I wasn't sure if it was a bug in AJV, in how
Spectral depends on AJV or how Spectral is actually using AJV.

But if it's definitely not a Spectral bug then I'll raise a parallel ticket
against AJV :)

Cheers

**Comment by skilbjo** [comment]
Hmm, I am seeing this with `npm@7`

I am using the https://github.com/IBM/openapi-validator library, and run its basic command, `lint-openapi openapi.yaml`, and get the exact same stack trace as @sazzer described above


```
$ npm --version
7.19.0
$ node --version
v14.17.0
```

```
$ npm ls ajv
api@0.0.1 /Users/skilbjo/dev/api
├─┬ ajv-formats@2.1.1
│ └── ajv@8.8.2 deduped
├── ajv@8.8.2
├─┬ eslint@8.1.0
│ ├─┬ @eslint/eslintrc@1.0.3
│ │ └── ajv@6.12.6
│ └── ajv@6.12.6
└─┬ ibm-openapi-validator@0.52.0
  ├─┬ @ibm-cloud/openapi-ruleset@0.1.0
  │ ├─┬ @stoplight/spectral-functions@1.5.0
  │ │ ├─┬ ajv-draft-04@1.0.0
  │ │ │ └── ajv@8.8.2 deduped
  │ │ └── ajv@8.8.2 deduped
  │ └─┬ @stoplight/spectral-rulesets@1.3.0
  │   ├─┬ @stoplight/better-ajv-errors@0.2.0
  │   │ └── ajv@8.6.3 deduped
  │   └── ajv@8.6.3
  ├─┬ @stoplight/spectral-cli@6.1.0
  │ └─┬ @stoplight/spectral-ruleset-migrator@1.6.0
  │   └── ajv@8.8.2 deduped
  └─┬ @stoplight/spectral-core@1.8.0
    ├─┬ @stoplight/better-ajv-errors@1.0.1
    │ └── ajv@8.8.2 deduped
    ├─┬ ajv-errors@3.0.0
    │ └── ajv@8.8.2 deduped
    └── ajv@8.6.3
```

```
> lint-openapi spec/openapi.yaml -e

Error compiling schema, function code: const schema20 = scope.schema[14];const schema21 = scope.schema[15];const schema22 = scope.schema[16];const func0 = scope.func[0];return function validate18(data, {instancePath="", parentData, parentDataProperty, rootData=data}={}){let vErrors = null;let errors = 0;const _errs0 = errors;let valid0 = false;let passing0 = null;const _errs1 = errors;if(!(((((data === -1) || (data === 0)) || (data === 1)) || (data === 2)) || (data === 3))){const err0 = {instancePath,schemaPath:"#/$defs/DiagnosticSeverity/enum",keyword:"enum",params:{allowedValues: schema21.enum},message:"must be equal to one of the allowed values"};if(vErrors === null){vErrors = [err0];}else {vErrors.push(err0);}errors++;}var _valid0 = _errs1 === errors;if(_valid0){valid0 = true;passing0 = 0;}const _errs3 = errors;if(!(((((data === "error") || (data === "warn")) || (data === "info")) || (data === "hint")) || (data === "off"))){const err1 = {instancePath,schemaPath:"#/$defs/HumanReadableSeverity/enum",keyword:"enum",params:{allowedValues: schema22.enum},message:"must be equal to one of the allowed values"};if(vErrors === null){vErrors = [err1];}else {vErrors.push(err1);}errors++;}var _valid0 = _errs3 === errors;if(_valid0 && valid0){valid0 = false;passing0 = [passing0, 1];}else {if(_valid0){valid0 = true;passing0 = 1;}}if(!valid0){const err2 = {instancePath,schemaPath:"#/oneOf",keyword:"oneOf",params:{passingSchemas: passing0},message:"must match exactly one schema in oneOf"};if(vErrors === null){vErrors = [err2];}else {vErrors.push(err2);}errors++;}else {errors = _errs0;if(vErrors !== null){if(_errs0){vErrors.length = _errs0;}else {vErrors = null;}}}if(errors > 0){for(const err3 of vErrors){if((((({"str":"err3"}.keyword !== "errorMessage") && (!{"str":"err3"}.emUsed)) && (({"str":"err3"}.instancePath === instancePath) || (({"str":"err3"}.instancePath.indexOf(instancePath) === 0) && ({"str":"err3"}.instancePath[instancePath.length] === "/")))) && ({"str":"err3"}.schemaPath.indexOf("#") === 0)) && ({"str":"err3"}.schemaPath["#".length] === "/")){{"str":"emErrs0"}.push({"str":"err3"});{"str":"err3"}.emUsed = true;}}if({"str":"emErrs0"}.length){if(vErrors === null){vErrors = [{"str":"err4"}];}else {vErrors.push({"str":"err4"});}errors++;}const emErrs1 = [];for(const err5 of vErrors){if(!{"str":"err5"}.emUsed){{"str":"emErrs1"}.push({"str":"err5"});}}vErrors = emErrs1;errors = {"str":"emErrs1"}.length;}validate18.errors = vErrors;return errors === 0;}
/Users/skilbjo/dev/api/node_modules/@stoplight/spectral-core/node_modules/ajv/dist/compile/index.js:121
        throw e;
        ^

SyntaxError: Unexpected token ':'
    at new Function (<anonymous>)
    at Ajv.compileSchema (/Users/skilbjo/dev/api/node_modules/@stoplight/spectral-core/node_modules/ajv/dist/compile/index.js:90:30)
    at Ajv.inlineOrCompile (/Users/skilbjo/dev/api/node_modules/@stoplight/spectral-core/node_modules/ajv/dist/compile/index.js:149:47)
    at Ajv.resolveRef (/Users/skilbjo/dev/api/node_modules/@stoplight/spectral-core/node_modules/ajv/dist/compile/index.js:143:46)
    at Object.code (/Users/skilbjo/dev/api/node_modules/@stoplight/spectral-core/node_modules/ajv/dist/vocabularies/core/ref.js:19:47)
    at keywordCode (/Users/skilbjo/dev/api/node_modules/@stoplight/spectral-core/node_modules/ajv/dist/compile/validate/index.js:454:13)
    at /Users/skilbjo/dev/api/node_modules/@stoplight/spectral-core/node_modules/ajv/dist/compile/validate/index.js:185:25
    at CodeGen.code (/Users/skilbjo/dev/api/node_modules/@stoplight/spectral-core/node_modules/ajv/dist/compile/codegen/index.js:439:13)
    at CodeGen.block (/Users/skilbjo/dev/api/node_modules/@stoplight/spectral-core/node_modules/ajv/dist/compile/codegen/index.js:568:18)
    at schemaKeywords (/Users/skilbjo/dev/api/node_modules/@stoplight/spectral-core/node_modules/ajv/dist/compile/validate/index.js:185:13)
```

**Comment by skilbjo** [comment]
Ah, looks like my issue and @sazzer 's issue is the same as discussed here: https://github.com/stoplightio/spectral/issues/1808 

and it appears currently unresolved: https://github.com/stoplightio/spectral/issues/1808#issuecomment-996698993

It is because of two versions of `ajv` (`6.x` for `eslint`, and `8.x` for modern `ajv`)
`eslint` discussed upgrading ajv to `8.x` here: https://github.com/eslint/eslint/issues/13888#issuecomment-872591875 but ultimately decided not to proceed.

Might there be any other workarounds for users who have both `eslint` and `spectral` in their dependencies?

**Comment by sazzer** [comment]
It's slightly more complex than that. It's also hampered by the fact that
Spectral itself uses two different versions - 8.6.x and 8.8.x.

I have managed to fix it by pinning the overall application to 8.6, which
then makes NPM put the other versions as child modules under the
appropriate areas. (6.x under ESLint and 8.8.x under the bits of Spectral
that use that) But this means that I've got a version of "8.6.3" instead of
"^8.6.3" in my package.json.

Putting the version "^8.6.3" into my package.json made NPM put 8.8.x in the
top-level, 8.6.x as a child under the bits of Spectral that used it, and
then that's when it all failed.

Cheers

**Comment by philsturgeon** [comment]
Fixed in #2020.

**Comment by DavidBiesack** [comment]
This is supposed to be fixed in #2020 but that is an update to spectral-rulesets (to 1.3.2 )
Like others, I'm using a tool which has its own CLI and it embeds Spectral directly (spectral-cli bundle, version 6.4.2)
and I still get the ajv error above

```
SyntaxError: Unexpected token ':'
```

due to my project using Spectral and also eslint and openapi-merge-cli which use ajv@6

```
@apiture/banking-api-def@0.44.0 ~/dev/*****
├─┬ @stoplight/spectral-cli@6.4.2
│ ├─┬ @stoplight/spectral-core@1.12.4
│ │ └── ajv@8.11.0 
│ └─┬ @stoplight/spectral-ruleset-migrator@1.7.3
│   └── ajv@8.11.0 
├─┬ @stoplight/spectral-rulesets@1.11.1
│ ├─┬ @stoplight/spectral-functions@1.6.1
│ │ └── ajv@8.11.0 
│ └── ajv@8.11.0 
├── ajv@8.11.0 
├─┬ ajv-cli@5.0.0
│ ├── ajv@8.11.0 
│ └─┬ json-schema-migrate@2.0.0
│   └── ajv@8.11.0 
├─┬ ajv-formats@2.1.1
│ └── ajv@8.11.0 
├─┬ eslint@8.20.0
│ ├─┬ @eslint/eslintrc@1.3.0
│ │ └── ajv@6.12.6 
│ └── ajv@6.12.6 
└─┬ openapi-merge-cli@1.3.1
  └── ajv@6.12.6 
```

Is there some way in my project (via npm config?) to have spectral pick up its own ajv@8 or even my project's explicit ajv@8 and not get ajv@6  ? I can perhaps get by w/o eslint but I can't remove [openapi-merge-cli](https://www.npmjs.com/package/openapi-merge-cli)
